cpp/src/strings/convert/convert_urls.cu

```diff
@@ -185,137 +192,184 @@ constexpr bool is_hex_digit(char ch)
-    size_type output_idx = input_idx - (num_prev_esc * 2) - first_input_char_offset;
-    if (num_prev_esc > 0) {
-      size_type prev_esc_pos = *(next_esc_pos_ptr - 1);
-      // find the previous escape to see if this character is within the escape sequence
-      if (input_idx - prev_esc_pos < 3) {
-        if (input_idx == prev_esc_pos) {  // at a position that needs to be replaced
-          ch = (hex_char_to_byte(d_in_chars[input_idx + 1]) << 4) |
-               hex_char_to_byte(d_in_chars[input_idx + 2]);
-          // previous escape sequence is this position, so the original calculation over-adjusted
-          output_idx += 2;
-        } else {
-          // one of the escape hex digits that has no corresponding character in the output
-          return;
-        }
+template <int threadblock_size, int block_size>
+__global__ void url_decode_char_replacer(char const* const in_chars,
+                                         offset_type const* const in_offsets,
+                                         char* const out_chars,
+                                         offset_type const* const out_offsets,
+                                         size_type const num_strings)
+{
+  constexpr int warps_per_threadblock = threadblock_size / cudf::detail::warp_size;
+  __shared__ char temporary_buffer[warps_per_threadblock][block_size + 4];
+  __shared__ typename cub::WarpScan<int8_t>::TempStorage cub_storage[warps_per_threadblock];
+  __shared__ int out_idx[warps_per_threadblock];
+
+  int global_thread_id  = blockIdx.x * blockDim.x + threadIdx.x;
+  int global_warp_id    = global_thread_id / cudf::detail::warp_size;
+  int local_warp_id     = threadIdx.x / cudf::detail::warp_size;
+  int warp_lane         = threadIdx.x % cudf::detail::warp_size;
+  int nwarps            = gridDim.x * blockDim.x / cudf::detail::warp_size;
+  char* in_chars_shared = temporary_buffer[local_warp_id];
+
+  // Loop through strings, and assign each string to a warp
+  for (size_type row_idx = global_warp_id; row_idx < num_strings; row_idx += nwarps) {
+    auto in_chars_string  = in_chars + in_offsets[row_idx];
+    auto out_chars_string = out_chars + out_offsets[row_idx];
+    auto string_length    = in_offsets[row_idx + 1] - in_offsets[row_idx];
+    int nblocks           = (string_length + block_size - 1) / block_size;
+
+    if (warp_lane == cudf::detail::warp_size - 1) { out_idx[local_warp_id] = 0; }
+
+    for (int block_idx = 0; block_idx < nblocks; block_idx++) {
+      int string_length_block = std::min(block_size, string_length - block_size * block_idx);
+
+      // Each warp collectively loads input characters of the current block to shared memory.
+      // Two halo cells with 0s before and after the input characters are added. The halo cells are
+      // used to test whether the current location as well as the previous two locations are escape
+      // characters, without branches.
```

Comment:
This comment is more informative than the one in the other kernel, thanks. However, why do you need to do a lookback in this case, but not in the counting case? In other words, why do you need to know whether the previous locations are escape characters here, but not in the counting case?

Reply:
An escaped sequence consists of three characters in the form of '%XX', where each 'X' is a hex digit. 

In the counting case, we would like to count the number of escaped sequences in each string. Each thread only needs to test whether its cell is the start location of a escaped sequence and then reduce. Note that if the current cell is the start location of an escaped sequence, this automatically means the previous two cells cannot be start locations of an escaped sequence, because '%' is not a hex digit.

In this copy kernel, suppose the current cell is not a start location of an escaped sequence, we still need to consider two cases.
1. If any of the previous two locations is the start of an escaped sequence, the current thread should do nothing, and the thread with the starting location should output the escaped character. I clarify this in https://github.com/rapidsai/cudf/pull/8622/commits/161468fdf6c4415d7990f887e8429c9e34800ec0.
2. If previous two locations are not start of an escaped sequence, we simply output the current character.

Reply:
Got it, I didn't realize that this is what `out_size == 1` was checking for, thanks.